---
title: "The UK Crop Microbiome Cryobank - mapping the files for the fastq checklist"
author: "Payton Yau"
date: "2024-01-23"
output:
  pdf_document: default
  html_document: default
---

# The UK Crop Microbiome Cryobank

The UK Crop Microbiome Cryobank integrates genomic (DNA) data with a cryobank collection of samples for the soil microbiomes of the UK major crop plant systems. For this project, the microbiomes are from the rhizosphere (the soil surrounding the crop plant roots) and from bulk soil (soil outside the rhizosphere). The Cryobank provides a facility for researchers to source data and samples, including cryo-preserved microbial material and genomic and metagenomic sequences from different soil microbiome environments.

## The script below was used for mapping the information for fastq checklist

Files required: 

1. The receipt after the plant checklist uploaded to the server

2. Plant checklist ERC000020 

3. MD5Checksum information

### Step 1

Load the table from the receipt `Webin` after the plant Checklist uploaded to the server

```{r Webin_load, warning=FALSE, message=FALSE}
Webin <- read.table("Webin-accessions-2023-12-07T15_42_52.222Z_OR.txt", header = T)

# Check the last 5 columns
tail(Webin)
```

```{r Webin_remove_last, warning=FALSE, message=FALSE}
# Remove the last column of the dataframe - `Webin`
Webin <- head(Webin, -1)

# Check the last 5 columns
tail(Webin)
```


### Step 2 

Load the data from the  Crop Check-list

```{r Crop_Check_list, warning=FALSE, message=FALSE}
Check.list <- read.delim('Checklist_GSC-MIxS_16Samplicons_OR_TESTv1.tsv', header = F, sep = "\t")

# Print the data (first 6 columns) to check if it has been read correctly
head(Check.list)
```
##### format the crop checklist to fit for the fastq checklist requirement and for the mapping/merging
```{r Crop_Check_list_mod, warning=FALSE, message=FALSE}
# Set the column names of 'data' to be the second row of 'data'
colnames(Check.list) <- Check.list[2,]

# Remove the first three rows of 'data'
Check.list <- Check.list[-c(1:3),]

# Create a new dataframe 'Check-list2' that only includes the 3rd and 4th columns of 'Check-list'
Check.list_2 <- Check.list[,c(3:4)]

# Print the data (first 6 columns) to check if it has been correctly
head(Check.list_2)
```

#### Merge 'Webin' and 'Check.list_2' on the columns "ALIAS" and "sample_alias", respectively, to create 'merged_df'
```{r merge_Webin_Check.list_2, warning=FALSE, message=FALSE}
merged_df <- merge(Webin, Check.list_2, by.x = "ALIAS", by.y = "sample_alias")

# Add a leading zero to all the numbers in the 'sample_title' column
merged_df$sample_title <- sub("-([0-9])([^0-9]|$)", "-0\\1\\2", merged_df$sample_title)

# Print the 'sample_title' column of 'merged_df' to check if it has been read correctly
print(merged_df$sample_title)
```

### Step 3 

Load the data from the "md5.txt"

```{r md5, warning=FALSE, message=FALSE}
md5 <- read.table("md5.txt", sep = "")

# Print the data (first 6 columns) to check the document structure
head(md5)
```

#### change the original md5 format to fit for fastq checklist
```{r md5_mod, warning=FALSE, message=FALSE}
# Keep only the first and second columns of 'md5'
md5 <- md5[, c("V2", "V1")]

# Create a new dataframe 'md5.R1' that only includes the rows of 'md5' where the second column contains "_R1_001.fastq.gz"
md5.R1 <- md5[grep("_R1_001.fastq.gz", md5$V2), ]

# Add a new column to 'md5.R1', which is created by splitting the second column on underscores and taking the first element
md5.R1$v3 <- sapply(strsplit(as.character(md5.R1$V2), "_"), `[`, 1)

# Replace single digits at the end of the strings in the new column with the same digit preceded by a zero
md5.R1$v3 <- sub("(\\d)$", "0\\1", md5.R1$v3)

# Replace single digits between dashes in the new column with the same digit preceded by a zero
md5.R1$v3 <- sub("-([1-9])-", "-0\\1-", md5.R1$v3)

# Set the column names of 'md5.R1'
colnames(md5.R1) <- c("forward_file_name","forward_file_md5", "sample_title")

# Create a new dataframe 'md5.R2' that only includes the rows of 'md5' where the second column contains "_R2_001.fastq.gz"
md5.R2 <- md5[grep("_R2_001.fastq.gz", md5$V2), ]

# Add a new column to 'md5.R2', which is created by splitting the second column on underscores and taking the first element
md5.R2$v3 <- sapply(strsplit(as.character(md5.R2$V2), "_"), `[`, 1)

# Replace single digits at the end of the strings in the new column with the same digit preceded by a zero
md5.R2$v3 <- sub("(\\d)$", "0\\1", md5.R2$v3)

# Replace single digits between dashes in the new column with the same digit preceded by a zero
md5.R2$v3 <- sub("-([1-9])-", "-0\\1-", md5.R2$v3)

# Set the column names of 'md5.R2'
colnames(md5.R2) <- c("reverse_file_name", "reverse_file_md5", "sample_title")

# Merge 'md5.R1' and 'md5.R2' on the "sample_title" column to create 'merged.md5'
merged.md5 <- merge(md5.R1, md5.R2, by="sample_title")

# Print the data (first 6 columns) to check the dataframe structure
head(merged.md5)
```

### Step 4

Merge "merged_df" and "merged.md5" on the "sample_title" column to create "merged_all"

```{r merged_all_formatting_1, warning=FALSE, message=FALSE}
merged_all <- merge(merged_df, merged.md5, by ="sample_title")

# Add several new columns to 'merged_all' with constant values
merged_all$study <- rep("PRJEB58189", nrow(merged_all))
merged_all$instrument_model <- rep("Illumina MiSeq", nrow(merged_all))
merged_all$library_name <- rep("Nextera XT v2", nrow(merged_all))
merged_all$library_source <- rep("METAGENOMIC", nrow(merged_all))
merged_all$library_selection <- rep("PCR", nrow(merged_all))
merged_all$library_strategy <- rep("AMPLICON", nrow(merged_all))
merged_all$library_layout <- rep("PAIRED", nrow(merged_all))

# Rename the "sample" column to "ACCESSION"
colnames(merged_all)[which(colnames(merged_all) == "sample")] <- "ACCESSION"

# Rename the "ACCESSION" column back to "sample"
colnames(merged_all)[colnames(merged_all) == 'ACCESSION'] <- 'sample'

# Reorder the columns of 'merged_all'
merged_all <- merged_all[, c("sample", "study", "instrument_model", "library_name", 
                             "library_source", "library_selection", "library_strategy", 
                             "library_layout", "forward_file_name", "forward_file_md5", 
                             "reverse_file_name", "reverse_file_md5")]

# Print the data (first 6 columns) to check the dataframe structure
head(merged_all)
```

### Step 5

Final modification for the fastq checklist

```{r merged_all_formatting_2, warning=FALSE, message=FALSE}
# Create a new row with the same number of columns as 'merged_all'
new_row <- setNames(data.frame(matrix(ncol = ncol(merged_all), nrow = 1)), colnames(merged_all))

# Assign the values to the new row
new_row[1, c("sample", "study", "instrument_model")] <- c("FileType", "fastq", "Read submission file type")

# Save the column names
col_names <- colnames(new_row) 

# Add the column names as a new row in the second position
new_row <- rbind(new_row, col_names)

# Add the new row to the top of the dataframe
merged_all <- rbind(new_row, merged_all)

# Remove column names
colnames(merged_all) <- NULL

# Print the data (first 6 columns) to check the dataframe structure
head(merged_all)
```

### Step 6

Export the table and ready to upload

```{r export, warning=FALSE, message=FALSE}
# write.table(merged_all, file = "fastq2_template_16Samplicons_OR_TEST_with_mapping.tsv", 
#             sep = "\t", row.names = FALSE, quote = FALSE, na = "")
```

```{r sessionInfo, warning=FALSE, message=FALSE}
sessionInfo()
```
